### DISCUSSION TITLE
Proposal: Conditional Compilation in XML Comments

### DISCUSSION THREAD

**Discussion by dotMorten** [dotnet/csharplang#295] [General]
Please support conditional compilation within XML documentation comments. This has been a source of pain with cross-compiled code for a long time. Often linked code files are used to share source between different target platforms of the same API; however, as a result often times the documentation comments must change conditionally as well. As of now, there's no good way around this limitation - you have to duplicate the entire XML comments conditionally, rather than just each section or line.

**Steps to Reproduce**:
1. Create the following property with doc in a class:

``` csharp
        /// <summary>
        /// Turns the GPS on or off
        /// </summary>
        /// <remarks>
        /// <para>Will turn on the system location service.</para>
        /// <para>Make sure you have enabled the location capability in your application's manifest. See
#if NETFX_CORE
        /// https://msdn.microsoft.com/en-us/library/windows/apps/br211423.aspx
#elif __ANDROID__
        /// http://developer.android.com/guide/topics/manifest/manifest-intro.html 
#endif
        /// for more information
        /// </para>
        /// </remarks>
        public bool IsGpsEnabled { get; set; }
```
1. Enable XML documentation generation 
2. Build

**Expected Behavior**:
No build warnings. XML is generated with platform specific remarks. This allows me to add a couple of platform specific remarks. Currently you have to if-def the entire code summary section, causing a lot of documentation to be duplicated, rather than just the areas that differ

**Actual Behavior**:
Several build warnings

```
warning CS1570: XML comment has badly formed XML -- 'Expected an end tag for element 'remarks'.'
1>\Class1.cs(25,13,25,14): warning CS1570: XML comment has badly formed XML -- 'End tag was not expected at this location.'
1>\Class1.cs(9,18,9,24): warning CS1591: Missing XML comment for publicly visible type or member 'Class1'
1>\Class1.cs(11,9,11,10): warning CS1587: XML comment is not placed on a valid language element
```

And the generated XML:

``` xml
<?xml version="1.0"?>
<doc>
    <assembly>
        <name>ClassLibrary1</name>
    </assembly>
    <members>
        <!-- Badly formed XML comment ignored for member "P:ClassLibrary1.Class1.IsGpsEnabled" -->
    </members>
</doc>
```

Moved over from: https://github.com/dotnet/roslyn/issues/96

**Comment by terrajobst** [comment]
@dotMorten 

Let's say instead of allowing `#if` in doc comments we'd extend the XML vocabulary to allow for expressing conditional sections, like so:

```C#
/// <summary>
/// Turns the GPS on or off
/// </summary>
/// <remarks>
/// <para>Will turn on the system location service.</para>
/// <para specificTo="NETFX_CORE">
/// Make sure you have enabled the location capability in your application's manifest. See
/// https://msdn.microsoft.com/en-us/library/windows/apps/br211423.aspx
/// for more information.
/// </para>
/// <para specificTo="__ANDROID__">
/// Make sure you have enabled the location capability in your application's manifest. See
/// http://developer.android.com/guide/topics/manifest/manifest-intro.html
/// for more information.
/// </para>
/// </remarks>
public bool IsGpsEnabled { get; set; }
```

In your example that seems overly verbose; but I think there might be a general need for marking parts of the documentation as only applying to a particular .NET platform or operating system. It would be nice if we could record this in the .XML file and allow the documentation tool to output them accordingly, i..e by placing them in specific sections with title or header or even making their visibility conditional on some context switcher on the web page.

Thoughts?

**Comment by jnm2** [comment]
@terrajobst I think keeping the current `#if` syntax is better. It would be more consistent with `#pragma warning disable`/`restore` lines in the XML documentation lines.

**Comment by nil4** [comment]
@terrajobst `#if` directives are documented to [support boolean operators, including `==`, `!=`, `&&`, `||` and `!`](https://msdn.microsoft.com/en-us/library/4y6tbswk.aspx), e.g. `#if (DEBUG && !MYTEST)`. It's unclear how the proposed `specificTo` attribute would cope with these and reduce duplication for more complex conditions.

**Comment by dotMorten** [comment]
@terrajobst Your approach won't work. We're compiling different binaries, which would then have different XML doc pieces. You would never need a "merged" xml doc, as the xml doc generated is specific to that binary. Second you would assume any consumer of that xml doc has the exact same If-Def defined. You would have to use the TFM instead, but TFM is quite different from if-defs which can be very custom (and most often are).

I think your suggestion comes from a point of view of .NET Standard usage. In that case, you wouldn't filter documentation, but make them different sections/paragraphs, so the user of the library would realize they need to do different things on different platforms with the same API.

**Comment by dotMorten** [comment]
An argument was made in the old thread that changing this was against the spec that states doc comments _have_ be to consecutive.

However I think the spec leaves plenty of room for interpretation here, and the way I read it, seems to support it's allowed.

> The conditional compilation directives are used to conditionally include or exclude portions of a source file.

If the section is excluded, that means the doc comments in fact does become consecutive.

> The selected conditional_section, if any, is processed as a normal input_section: the source code contained in the section must adhere to the lexical grammar; tokens are generated from the source code in the section; and pre-processing directives in the section have the prescribed effects.

and

> The remaining conditional_sections, if any, are processed as skipped_sections: except for pre-processing directives, the source code in the section need not adhere to the lexical grammar; 

After excluding/including the section, the source code now adheres to the lexical grammar.

These directives are called _pre_ processing for a reason. If you pre-process the code, then the documentation is in fact following the spec.

**Comment by FransBouma** [comment]
I ran into this today and it's odd that this happens at all. I'm with @dotMorten in that pre-processing should produce the actual lexical source processed by the lexer/parser and it's surprising that this doesn't work this way today.

Any news on this (but as this issue is rather old, I guess: nope) ?

**Comment by terrajobst** [comment]
@dotMorten 

> You would never need a "merged" xml doc, as the xml doc generated is specific to that binary. 

I was thinking about how we'd surface `#if` in documentation experiences that are produced from the .xml files (e.g. docs.microsoft.com). I guess considering that we basically don't use tripple slash comments as the source of truth for online docs it doesn't really matter.

**Comment by jnm2** [comment]
I had a hard time believing it till I saw that almost-verbatim copies is the only way to slightly modify documentation across target frameworks:

```cs
#if NETSTANDARD1_3
        /// <summary>
        /// <para>
        /// Selects the ultimate shadowing property just like <see langword="dynamic"/> would,
        /// rather than throwing <see cref="AmbiguousMatchException"/>
        /// for properties that shadow properties of a different property type.
        /// </para>
        /// <para>
        /// If you request both public and nonpublic properties, every public property is preferred
        /// over every nonpublic property. It would violate the principle of least surprise for a
        /// derived class’s implementation detail to be chosen over the public API for a type.
        /// </para>
        /// </summary>
#elif NETSTANDARD1_6
        /// <summary>
        /// <para>
        /// Selects the ultimate shadowing property just like <see langword="dynamic"/> would,
        /// rather than throwing <see cref="AmbiguousMatchException"/>
        /// for properties that shadow properties of a different property type
        /// which is what <see cref="TypeInfo.GetProperty(string, BindingFlags)"/> does.
        /// </para>
        /// <para>
        /// If you request both public and nonpublic properties, every public property is preferred
        /// over every nonpublic property. It would violate the principle of least surprise for a
        /// derived class’s implementation detail to be chosen over the public API for a type.
        /// </para>
        /// </summary>
        /// <param name="type">See <see cref="TypeInfo.GetProperty(string, BindingFlags)"/>.</param>
        /// <param name="name">See <see cref="TypeInfo.GetProperty(string, BindingFlags)"/>.</param>
        /// <param name="bindingFlags">See <see cref="TypeInfo.GetProperty(string, BindingFlags)"/>.</param>
#else
        /// <summary>
        /// <para>
        /// Selects the ultimate shadowing property just like <see langword="dynamic"/> would,
        /// rather than throwing <see cref="AmbiguousMatchException"/>
        /// for properties that shadow properties of a different property type
        /// which is what <see cref="Type.GetProperty(string, BindingFlags)"/> does.
        /// </para>
        /// <para>
        /// If you request both public and nonpublic properties, every public property is preferred
        /// over every nonpublic property. It would violate the principle of least surprise for a
        /// derived class’s implementation detail to be chosen over the public API for a type.
        /// </para>
        /// </summary>
        /// <param name="type">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
        /// <param name="name">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
        /// <param name="bindingFlags">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
#endif
```

https://github.com/nunit/nunit/blob/111fc6b5550f33b4fceb6ac8693c5692e99a5747/src/NUnitFramework/framework/Internal/Reflect.cs#L230-L277

**Comment by CyrusNajmabadi** [comment]
> However I think the spec leaves plenty of room for interpretation here, and the way I read it, seems to support it's allowed.

> These directives are called pre processing for a reason. If you pre-process the code, then the documentation is in fact following the spec.

The thing is, it's not really 'pre processing'.  It's in-situ processing.  That's why, for example, you can write the following:

```c#
Console.WriteLine(@"

#if false
This is all inside the string and will get printed to the console
#endif

");
```

If this truly were preprocessed, then the part between "#if false" and "#endif" would be entirely removed.  

Now, that said, i don't see any real problem with the compiler taking the sections of trivia that are doc comments, removing the disabled text sections, and doing final analysis as if it was one contiguous region of doc comments.

**Reply by Patrick8639** [reply]
This is not exactly the same thing, as an XML comment ends at each line, so there is no problem with the code in the original post. Enabling #if in whole lines should be enough.

**Comment by jnm2** [comment]
> Seems like you're the first to ever ask for it :)

I'm sorry, I didn't mean that to sound harsh. @dotMorten says at the top this has been a source of pain for people for a long time. It just got real for me.

This should be easy to approve and then super easy to implement in roslyn, shouldn't it?

**Comment by jnm2** [comment]
> ```cs
> Console.WriteLine(@"
> 
> #if false
> This is all inside the string and will get printed to the console
> #endif
> 
> ");
> ```
> 
> If this truly were preprocessed, then the part between "#if false" and "#endif" would be entirely removed.

This just blew my mind. 😃

**Reply by Patrick8639** [reply]
This is not exactly the same thing, as an XML comment ends at each line, so there is no problem with the code in the original post. Enabling #if in whole lines should be enough.

**Comment by CyrusNajmabadi** [comment]
> This should be easy to approve and then super easy to implement in roslyn, shouldn't it?

There would certainly be complexity.  Enough that i would not call it "super easy to implement".  For example, the lexer would have to have pretty interesting changes to it.  Consider the following simple case:

```c#
/// <summary>
#if false
...
#endif
/// </summary>
```

Recall that *currently* parsing of doc comments happens in situ with *trivia* scanning.  i.e. while you're in the middle of a scan, and you are handling trivia, we recursively go into parsing XML for the doc comment.  THis parse would then have to essentially pause and store state while it recursively handled the ```#if ar || bi && tra (ry) || (sub && expression)``` section.  Then, once it got out of the #if/#elif#else/#endif region it would have to pop back it's doc comment parsing state to continue.

Certainly doable.  But also something that would take methodical work to ensure that no bugs/strangeness what introduced while this happened.

This would also certainly change the roslyn parse tree model (a breaking change which we would have to accept).

Today, the above xml snippet would be three distinct trivia sections.  Now, you'd want it to be one section (so you could get one single structured trivia node out).   In that sub-parse-tree, the ```#if``` section would then be nested trivia inside of it.  

We would also have to test/validate all IDE code to make sure it could handle these new trees.

So, i would place this at medium cost+complexity to implement.  Certainly not impossible.  But definitely not a 5 minute fix either :)

**Comment by dotMorten** [comment]
Only medium? Let's do it! 😁

@jnm2 your example does actually work. The trick is to if-def the ENTIRE XML doc section. However that usually means lots of duplicate XML doc to maintain when often it's just a single paragraph or less that needs to be tweaked for each platform.

**Comment by jnm2** [comment]
@dotMorten Yep, the example was what I was complaining about. 😃

**Comment by Patrick8639** [comment]
Here is an open-source tool I wrote to post-process the XML comments: https://github.com/Patrick8639/ConditionalXmlComments
HTH

**Comment by Poltuu** [comment]
any news on this ?

**Reply by Patrick8639** [reply]
Poltuu,
In between, you can use my tool (see the post). It's not perfect, but it makes the job done in my case.

**Reply by 333fred** [reply]
There is no news on this.

**Comment by rdeago** [comment]
I just ran into the same problem. Nice to see I'm not alone; less nice to see there's too few of us, and it happens too infrequently, to make a dent in the roadmap. I guess I'll live with it. 😐

**Comment by TaffarelJr** [comment]
Same. I'm multi-targeting several platforms at once, each with slightly different XML comments (references to .NET components, links to documentation, etc). This means many ~20-line comment blocks, each duplicated 3 or 4 times.

Super frustrating, very noisy, and prone to dulication errors. Seems like the intuitive use of PP directives in XML comments ought to be something that is supported.

**Reply by Patrick8639** [reply]
@TaffarelJr Did you see my post from November 18, 2021?
I have made a tool to address this problem. It's not perfect, but it does the job, at least for me.

**Comment by terrajobst** [comment]
The .NET platform teams has moved to using triple-slash (documentation comments) as the source of truth for most of the class libraries. I believe this is what most developers do.

The scenario here is multi-targeted builds where a single project produces builds for multiple targets (usually different frameworks) and therefore produces multiple binaries, and, by extension, multiple XML documentation files. In the context of IntelliSense (code completion in an IDE that shows these comments as tooltips) having multiple files is pretty easy to reason about.

My question is how we'd use this model for generating documentation that is meant to cover a single multi-targeted library. Like a web experience like docs.microsoft.com.

One could envisage a framework switcher, akin to the toggle in the editor, but I don't think this is desirable. I believe developers would want a single document for a given API provided by a library where framework/platform differences are explained in text, preferably with some visual marker. My rationale is that if you build cross-platform code consumers often need to reason about how an API works across multiple targets. Not every consumer is an application author that only targets a single framework/operating system.

One way to do this is to leave it up to the documentation tool. In a sense, that's already what is happening because each binary produced by a multi-targeted build can have a different API surface. Therefore, we generally expect a documentation generator to merge all APIs into a single table of contents for instance. The question is how we'd deal with disjoint contents of the XML documentation. I guess a tool could merge them on a per-XML node basis and add metadata which leg of the multi-targeted build this is specific too.

I think it would be worthwhile to explore a few examples to see how one would want to author it on the C# side and how one would want to see it appear on the docs. If we take the example above I can see it convenient on the C# side but when I read the docs I would probably want to see more like this:

> Make sure you have enabled the location capability in your application's manifest. The mechanism by which you do this differs by operating system:
> * On **Android**, see [App manifest](http://developer.android.com/guide/topics/manifest/manifest-intro.html).
> * On **Windows**, see [AppX manifest](https://msdn.microsoft.com/en-us/library/windows/apps/br211423.aspx).

I'm not convinced I'd want to use conditional compilation on the C# side to model this, but I do think I'd want some way to express this in the documentation.

**Reply by dotMorten** [reply]
This feels like quite an opinionated view of how doc should be. Wrt to TFM-switcher [this is exactly what we do](https://developers.arcgis.com/net/api-reference/api/netwin/Esri.ArcGISRuntime/Esri.ArcGISRuntime.Location.NmeaLocationDataSource.html). Some platforms might have more APIs than other platforms, because it doesn't always apply to all platforms, or there are platform specific APIs in some builds (one example is anything that works with bluetooth, where Windows, Android and iOS all take different native platform-specific BT-devices as parameters, but where the class description thus always can become rather platform-specific and refer to different method signatures in their cref links - cref links that aren't valid in all TFMs and would cause build warnings and docfx errors).
It's also not how DocFX is set up to build as it doesn't generate a merged doc, but one for each TFM.

**Reply by terrajobst** [reply]
> This feels like quite an opinionated view of how doc should be.

It totally is :-) I'm not mean to dictate a specific outcome here, by the way. Multi-targeting is more of an advanced scenario and not many developers have end-to-end that includes documentation yet, us included. Hence, I'm asking the question what we want the outcome to be.

**Reply by dotMorten** [reply]
And I don't disagree with you that for crossplatform APIs it is nice to have all platform behavior differences documented for all TFMs - it's just that there are technical reasons like you'd still need to be able to conditionalize because cref doesn't work for all platforms if all platforms doesn't have the same set of members.

**Reply by Patrick8639** [reply]
In my case, I have common code between the web service and the client side, the difference being that one side talks with an SQL server and processes requests from the client, and the other side sends requests to the web service. So the documentation must sometimes be different between the two parts.

**Comment by DS-AdamMilazzo-CP** [comment]
It's not perfect, but you can move some of your documentation into an external file and `<include>` it, something like

```C#
#if FOO
/// <include file="documentation.xml" path="/MyLib/MyClass/MyFunc/*[not(self::remarks)]" />
/// <remarks><include file="documentation.xml" path="/MyLib/MyClass/MyFunc/remarks/*[not(@condition) or @condition='foo']" /></remarks>
#else
/// <include file="documentation.xml" path="/MyLib/MyClass/MyFunc/*[not(self::remarks)]" />
/// <remarks><include file="documentation.xml" path="/MyLib/MyClass/MyFunc/remarks/*[not(@condition)]" /></remarks>
#endif
```

This minimizes the duplication at the cost of having to edit some of your documentation in an external file.

For what it's worth, I also wish `#if` worked with XML doc comments.

**Reply by Patrick8639** [reply]
@DS-AdamMilazzo-CP Did you see my post from November 18, 2021?
I have made a tool to address this problem. It's not perfect, but it does the job, at least for me.

**Reply by DS-AdamMilazzo-CP** [reply]
I saw it. It's nice and simple but it doesn't seem like the kind of thing that will just work "out of the box" if somebody pulls code with conditional XML and tries to build it. They'd also have to pull your tool, build it, and either run it manually or put it somewhere that the post-build event will find it.

I don't know if it's possible, but it'd be better if it was something like a Nuget package that you could reference. Then it would be transparent and not require additional setup for somebody to compile the code. Something like a Roslyn code decorator or whatever the proper terminology is.

**Comment by jrmoreno1** [comment]
@dotMorten : Not sure if you meant source generator or not, but I would think it would be possible with one.
